README.md
# EPP <img align="right" src="man/figures/epp_logo_b.png" alt="logo" width="160"> <img align="right" src="man/figures/epp_logo.png" alt="logo" width="160">

<!-- badges: start -->
  [![CRAN status](https://www.r-pkg.org/badges/version/EPP)](https://CRAN.R-project.org/package=EPP)
  [![CRAN/METACRAN Total downloads](http://cranlogs.r-pkg.org/badges/grand-total/EPP?color=blue)](https://CRAN.R-project.org/package=EPP) 
  [![CRAN/METACRAN downloads per month](https://cranlogs.r-pkg.org/badges/EPP?color=blue)](https://CRAN.R-project.org/package=EPP)
  [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.5138170.svg)](https://doi.org/10.5281/zenodo.5138170)
  [![R build status](https://github.com/RichDeto/EPP/workflows/R-CMD-check/badge.svg)](https://github.com/RichDeto/EPP/actions)
  [![codecov](https://codecov.io/gh/RichDeto/EPP/branch/master/graph/badge.svg)](https://app.codecov.io/gh/RichDeto/EPP)
  [![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)

[![AppVeyor build status](https://ci.appveyor.com/api/projects/status/github/RichDeto/EPP?branch=master&svg=true)](https://ci.appveyor.com/project/RichDeto/EPP)
<!-- badges: end -->

## Evaluation of Proximity Programs

**EPP** is a library oriented to the evaluation of population coverage of diverse programs. 
It was originally developed to assess the coverage of an initial education program in Uruguay, which gave us interesting results, and we wanted to share the methodology used, in a R package format. 
All contributions are welcome, even though we are still in the process of improvement.

## Installation

```R
# From CRAN

  install.packages("EPP")
  library(EPP)
```

```R
# Use the development version with latest features

  utils::remove.packages('EPP')
  devtools::install_github("RichDeto/EPP")
  library(EPP)
```

If you use **Linux**, you need to install a couple dependencies before installing the libraries `{sf}` and `{EPP}`. [More info here](https://github.com/r-spatial/sf#linux). 

## Basic Usage

### Main functions

A first task is to estimate the population covered with the existing infrastructure, and therefore we refer to the `eppexist()` function. An important detail to take into account is that our centers cannot overlap, since this prevents the correct computation of the Voronoi polygons. For this particular case, the `group_over()` function was implemented that combines the records for this specific application case.

Using the datasets `pop_epp` and `centers_epp` of the own library, this is it use. 

```R
centers_epp <- group_over(rbind(centers_epp, centers_epp[ 1:3,]))
exist <- eppexist(pop = pop_epp, centers = centers_epp, crs = sp::CRS("+init=epsg:32721"))
```

:warning: If you need to process a large number of cases with the parameter "route" = TRUE, it is recommended to install OSRM on a local server. For more information take a look [here](https://github.com/riatelab/osrm/issues/4) and [here](https://github.com/Project-OSRM/osrm-backend/issues/5463).

Normally the population is not completely covered by existing infrastructure, and that's when `eppproy()` appears, a function to find optimal locations to create new centers to cover the remaining population.  

Continuing with the example:

```R
proy <- eppproy(pop = exist$pop_uncover)
```

This was just an example using the default values of all the parameters. Please play around with them and report any bug [here](https://github.com/RichDeto/EPP/issues/new/choose).

### Report template

The library also has templates in English and Spanish to quickly produce report of the results of the processing.
In RStudio when you create a new RMarkdown document you can select the template like image show and start to personalize it. 

<img src="man/figures/template_epp.PNG" width="50%" style="display: block; margin: auto;" />

### Visualizations

For a quick visualization of the results of the `eppexist` or `eppproy` functions, the `leafepp` function was provided, which generates a `{leaflet}` viewer, with all the sublayers of the process. It also has a version in English and Spanish.

Continuing with the example:

```R
## In case of eppexist (In English)
l_epp_exist <- leafepp(exist, t = "exist", crs = sp::CRS("+init=epsg:32721"), leng = "en")
l_epp_exist
```

```R
## In case of eppproy (and in Spanish)
l_epp_proy <- leafepp(proy, t = "proy", crs = sp::CRS("+init=epsg:32721"), leng = "es")
l_epp_proy
```

### General functions

The syntax of all `{EPP}` functions are focused on executing two main processes, `eppexist()` and `eppproy()`, both aimed at evaluating the distribution of a certain population and the centers planned to cover it. Under the hood, there are some other tools that can be useful for other processes. Among them, we can mention those that allow making voronoi polygons (`voro_polygon()`), buffer-voronoi (`buff_voronoi()`) and iterative clusters (`clust_it()`).

## References

Detomasi, R. 2018. "Abordaje espacial de políticas públicas: cuidados y primera infancia”. En: [Las políticas públicas dirigidas a la infancia en Uruguay](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=2&cad=rja&uact=8&ved=2ahUKEwiu_uH64vHjAhXsJ7kGHRHOCo4QFjABegQIABAC&url=https%3A%2F%2Frepositorio.cepal.org%2Fbitstream%2Fhandle%2F11362%2F44155%2F1%2FS1800463_es.pdf&usg=AOvVaw3EPJkSZSWIDsQ-dpwcHuUO). Coords. Verónica Amarante  y Juan Pablo Labat. Ed. CEPAL - UNICEF, Santiago de Chile.

Detomasi, R., G. Mathieu y G. Botto 2018. ["EPP v.0.2: Evaluation of Proximity Programs with OSRM routing"](https://47jaiio.sadio.org.ar/sites/default/files/LatinR_10.pdf). LatinR - Conferencia Latinoamericana sobre Uso de R en Investigación + Desarrollo. 3 al 7 de Setiembre. 

Detomasi, R. y G. Botto. 2017. ["Evaluación espacial de servicios de educación inicial: la densificación de la oferta para tres años en la Administración Nacional de Educación Pública (ANEP)"](https://www.researchgate.net/publication/322149061_EVALUACION_ESPACIAL_DE_SERVICIOS_DE_EDUCACION_INICIAL_LA_DENSIFICACION_DE_LA_OFERTA_PARA_NINOS_DE_TRES_ANOS_EN_LA_ADMINISTRACION_NACIONAL_DE_EDUCACION_PUBLICA_ANEP). GeoFocus (ISSN 1578-5157).
<!-- https://www.geofocus.org/index.php/geofocus/article/view/508). -->

Botto, G. y Detomasi, R. 2015. ["Bases metodológicas para la planificación espacial de servicios de educación inicial en Uruguay"](https://dinem.mides.gub.uy/innovaportal/file/61794/1/tecnologias-de-la-informacion-para-nuevas-formas-de-gestion-del-territorio.-2015.pdf) Jornadas Argentinas de Geo-tecnologías: Trabajos completos. Universidad Nacional de Luján - Sociedad de Especialistas Latinoamericanos en Percepción Remota - Universidad Nacional de San Luis, pp. 121-128.

Detomasi, R., Botto, G. y Hahn, M. 2015. ["CAIF: Análisis de demanda"](https://dinem.mides.gub.uy/innovaportal/file/61792/1/caif.-analisis-de-demanda.-2015.pdf) Documento de trabajo, Mayo 2015. Departamento de Geografía. Dirección Nacional de Evaluación y Monitoreo. Ministerio de Desarrollo Social.  

R Development Core Team 2015. ["R: A language and environment for statistical computing"](https://www.R-project.org/) R Foundation for Statistical Computing, Vienna, Austria.ISBN 3-900051-07-0.


NEWS.md
# log history of EPP package development

-------------------------------------------------------

## EPP v0.3.6.2 (2022-05-01)

  * Add osrm_ok() to others functions to fail gracefully with an informative message

## EPP v0.3.6.1 (2022-01-24)

  * update dependencie to osrm (>= 3.5.1)

## EPP v0.3.6 (2022-01-23)

  * Add methods::is function in assign_nn function to stop failling in Fedora
  * Demove ‘clipr’ ‘cpp11’ ‘s2’ ‘tibble’ as dependencies
  
## EPP v0.3.5 (2021-12-10)

  * Delete 'as.vector()' in three functions to stop failling in latest  R-devel

## EPP v0.3.4 (2021-11-04)

  * added a osrm_ok function to fail gratefully if osrm server is down
  * fixed 'F' abreviation in eppexist
  * added a return to osrm_ok
  * replace '\dontrun{}' with '\donttest{}' in osrm functions.

## EPP v0.3.3 (2021-09-22) 

  * leafepp debuged in t = "exist" about center$capacity popup
  * debuged assign_clust line 15
  * categorization of centers by coverage in leafepp(t = "proy")
  * added assign_nn() function
  * added addSearchOSM() in leafepp
  * added labels_es dataset to translate leafepp maps 
  * adaptation to osrm update
  * added site by pkgdown
  * added assertthat to check internet availability
  * debuged typo in assign_clust of p_weight

## EPP v0.3.2 (2021-08-03)

Launch of **EPP** v0.3.2 on [GitHUb](https://github.com/RichDeto/EPP) with the following functions:  
  * `assign_clust()`    
  * `assign_voro()`    
  * `assignation_exist()`
  * `buff_voronoi()`
  * `byosrmRoute()`
  * `clust_it()`
  * `clust_pop()`
  * `eppexist()`
  * `eppproy()`
  * `group_over()`
  * `leafepp()`
  * `osrm_matrixby100()`
  * `voro_polygon()`
  
  
And the following data sets:    
  * centers_epp    
  * pop_epp    


R/eppexist.R
#' Evaluation of proximity coverage by a program and estimate new services to attend the uncover population.
#' @description Assigns the elements from pob to the centers by n+m iterations, according to the capacity and distance/s of radium of each center. The output is a list that contains four data.frames: pop have the unassigned population, centers has the centers that still have capacity available, used_capacity have the center's occuped capacity and assigned has the population assigned to the centers. The assignation of points to centers is based in proximity and optimization criteria (maximum distance and center's capacity).
#' 
#' @param pop Population to attend (dataframe with three variables: x, y, and weight). Each element represents the position and identity of a population's member. x and y represents plain coordinates in the selected Coordinate Reference System.
#' @param centers Existing centers of the program (dataframe with four variables: x, y, id, capacity). Each element represents the position, identity and capacity of a center. x and y represents plain coordinates in the selected Coordinate Reference System.
#' @param n Total number of iterations with the distance "d1". Default n = 3
#' @param m Number of iteration in which change the distance of radium. If m>n only the first distance is used. Default m = 0
#' @param d1 Radius in meters that each center covers in the firsts "n" iterations. Default d1 = 1000
#' @param d2 Radius in meters that each center covers, the last "m" iterations. Default = d1 * 2
#' @param crs Coordinate Reference Systems (CRS). Default = CRS("+init=epsg:32721").
#' @param route logical if FALSE the distance is calculated by Pythagorean formula, if TRUE the distance is calculated by "osrmRoute" function of "osrm" Package. Default = FALSE
#'
#' @return Return a LIST with:
#' \item{pop_uncover }{DataFrame of the population still out of coverage; with its "x", "y" and "weight".}
#' \item{pop_assigned }{DataFrame with the population assigned and the corresponding center; with its "x", "y","weight", "id" of the center and the "iteration" of assign.}
#' \item{remaining_capacity }{DataFrame of the centers and its unused capacity; with "x" and "y" of the center, the "id", and the unused "capacity".}       
#' \item{used_capacity }{List of DataFrames with the centers and the info of population covered in each iteration. Each DataFrames have the name of the iteration, and contains the "id", "x" and "y" of the center; and the remaining "capacity" and "used_cap" after the iteration.}
#' @export
#' @references Bivand, R., Keitt, T., Rowlingson, B., Pebesma, E., Sumner, M., Hijmans, R. y Rouault, E. (2015), Bindings for the Geospatial Data Abstraction Library. URL https://cran.r-project.org/web/packages/rgdal/rgdal.pdf
#' Botto, G. y Detomasi, R. (2015), Bases metodológicas para la planificación espacial de servicios de educación inicial en Uruguay. Jornadas Argentinas de Geo-tecnologías: Trabajos completos. Universidad Nacional de Luján - Sociedad de Especialistas Latinoamericanos en Percepción Remota - Universidad Nacional de San Luis, pp. 121-128.http://dinem.mides.gub.uy/innovaportal/file/61794/1/tecnologias-de-la-informacion-para-nuevas-formas-de-gestion-del-territorio.-2015.pdf
#' Detomasi, R., Botto, G. y Hahn, M. (2015) CAIF: Análisis de demanda. Documento de trabajo, Mayo 2015. Departamento de Geografía. Dirección Nacional de Evaluación y Monitoreo. Ministerio de Desarrollo Social. http://dinem.mides.gub.uy/innovaportal/file/61792/1/caif.-analisis-de-demanda.-2015.pdf 
#' Farmer, carson (2009) Voronoi polygons with R. URL http://carsonfarmer.com/2009/09/voronoi-polygons-with-r/
#' Leisch, F. y Dimitriadou, E. (2013), Flexible Cluster Algorithms. URL https://cran.r-project.org/web/packages/flexclust/flexclust.pdf
#' Pebesma, E., Bivand, R., Rowlingson, B., Gomez-Rubio, V., Hijmans, R., Sumner, M., MacQueen, D. et al. (2015), Classes and Methods for Spatial Data. URL https://cran.r-project.org/web/packages/sp/sp.pdf
#' R Development Core Team (2015), R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria.ISBN 3-900051-07-0, URL http://www.R-project.org/
#' Turner, R. (2015), Delaunay Triangulation and Dirichlet (Voronoi) Tessellation. URL https://cran.r-project.org/web/packages/deldir/deldir.pdf
#' 
#' @examples
#' exist <- eppexist(pop = pop_epp, 
#'                   centers = centers_epp)
 
eppexist <- function(pop, centers, n = 3, m = 0, d1 = 1000, d2 = d1 * 2, crs = sp::CRS("+init=epsg:32721"), route = FALSE) {
  assigned <- as.list(NA)
  used_capacity <- as.list(NA) 
  dist <- c(rep.int(d1, n), rep.int(d2, m)) ## compiles a vector of distances
  for (i in 1:length(dist)) {
    if (nrow(pop) > 0 & nrow(centers) > 0) {
      iteration <- assignation_exist(pop = pop, centers = centers, d = dist[i], crs = crs, route = route) ## assigns the population to the centers
      pop <- iteration[[1]]
      cov <- iteration[[2]]
      if (nrow(cov) > 0) {
        cov$it <- i 
        cov$dist <- dist[i]
        cov$one <- 1
        cent1 <- merge(centers,
                    data.frame(used_cap = tapply(X = cov$one, INDEX = list(cov$id), FUN = sum)),
                    by.x = "id", by.y = "row.names", all.x = TRUE)
        cent1$used_cap <- ifelse(is.na(cent1$used_cap), 0, cent1$used_cap)
        cent1$capacity <- cent1$capacity - cent1$used_cap
        centers <- subset(cent1, cent1$capacity > 0, select = c(x, y, id, capacity), drop = TRUE) 
        centers <- droplevels(centers)
        used_capacity[[i]] <- cent1
        assigned[[i]] <- cov
        remove(cent1)
      } 
    remove(cov)
    }
  }
  names(used_capacity) <- 1:length(used_capacity)
  list("pop_uncover" = pop, "pop_assigned" = do.call("rbind", assigned)[ ,1:6], "remaining_capacity" = centers, 
       "used_capacity" = used_capacity[!sapply(used_capacity, is.null)])
}
